packages/atlas-service/src/main.ts

```diff
@@ -210,67 +348,65 @@ export class AtlasService {
+    // logged a refresh-succeeded event and this listener got triggered
     if (this.refreshing) {
       return;
     }
     this.refreshing = true;
+    const onRefreshFailed = () => {
+      this.token = null;
+      this.oidcPluginSyncedFromLoggerState = 'unauthenticated';
+      this.oidcPluginLogger.emit('atlas-service-token-refresh-failed');
+    };
     log.info(
       mongoLogId(1_001_000_214),
       'AtlasService',
       'Start atlas service token refresh'
     );
     try {
-      // We expect only one promise below to resolve, to clean up listeners that
-      // never fired we are setting up an abort controller
-      const listenerController = new AbortController();
-      try {
-        await Promise.race([
-          // When oidc-plugin logged that token was refreshed, the token is not
-          // actually refreshed yet in the plugin state and so calling `REFRESH_TOKEN_CALLBACK`
-          // causes weird behavior that actually opens the browser again, to work
-          // around that we wait for the state update event in addition. We can't
-          // guarantee that this event will be emitted for our particular state as
-          // this is not something oidc-plugin exposes, but we can ignore this for
-          // now as only one auth state is created in this instance of oidc-plugin
-          once(this.oidcPluginLogger, 'mongodb-oidc-plugin:state-updated', {
-            signal: listenerController.signal,
-          }),
-          // At the same time refresh can still fail at this stage, so to avoid
-          // refresh being stuck, we also wait for refresh-failed event and throw
-          // if it happens to avoid calling `REFRESH_TOKEN_CALLBACK`
-          once(this.oidcPluginLogger, 'mongodb-oidc-plugin:refresh-failed', {
-            signal: listenerController.signal,
-          }).then(() => {
-            throw new Error('Refresh failed');
-          }),
-        ]);
-      } finally {
-        listenerController.abort();
+      if (waitForOIDCPluginStateUpdateEvents) {
+        // We expect only one promise below to resolve, to clean up listeners
+        // that never fired we are setting up an abort controller
+        const listenerController = new AbortController();
+        try {
+          // When oidc-plugin starts internal token refresh, we will wait
+          // first for the auth flow to either succeed to fail and then
+          // proceed by requesting token once again from the plugin to sync
+          // token back to the atlas service state
+          await Promise.race([
+            // We are not using `refresh-succeeded` / `refresh-failed` events
+            // here because those happen BEFORE `allowedFlows` method is called
+            // (see oidc-plugin flow diagram) meaning that we need to wait until
+            // the whole request token flow went through to make sure that sign
+            // in flow is not allowed while we are refreshing the token
```

Comment:
That's the main change from the previous implementation: to make sure we don't return any `allowedFlows` before they are read and can be used on refresh failing we are listening to `auth-*` events here now instead of `refresh-*`, even though we are waiting specifically for refresh to finish